File: 3.2 Complete Search/Iterative 3D Harder/p11236.cpp
// Competitive Programming 3
// Problem 11236

#include <iostream>
using namespace std;

int main(){
    int target = 2000;

    for (int a = 1; a <= target-3; a++){
        for (int b = 1; b <= min(target-a-2, a); b++){
            for (int c = 1; c <= min(target-a-b-1, b); c++){
                int d;

                if (a*b*c <= 1000000)
                    continue;

                if ((1000000 * (a+b+c)) % (a*b*c - 1000000) == 0)
                    d = (1000000 * (a+b+c)) / (a*b*c - 1000000);
                else
                    continue;

                if (d > c || a+b+c+d > 2000)
                    continue;

                if (a * b * c * d == 1000000 * (a + b + c + d)){
                    printf("%d.%.2d %d.%.2d ", d/100,d%100,c/100,c%100);
                    printf("%d.%.2d %d.%.2d\n", b/100,b%100,a/100,a%100);
                }
            }
        }
    }
    return 0;
}

File: 3.2 Complete Search/Iterative 3 (Harder)/p11236.py
# Competitive Programming 3
# Problem 11236

target = 2000

# only search for solutions a >= b >= c >= d to avoid duplication

for a in range(1,target-2): 
    for b in range(1,min(target-a-1, a+1)):
        for c in range(1,min(target-a-b-1, b+1)):
            
            if a*b*c <= 100**3:
                continue
            if (100**3 * (a+b+c)) % (a*b*c - 100**3) == 0 :
                
                d = (100**3 * (a+b+c)) // (a*b*c - 100**3)
            
            if d > c or a+b+c+d > 2000: continue
            
            if a * b * c * d == 100**3 * (a + b + c + d):
                print("{0:.2f} ".format(d/100), end='')
                print("{0:.2f} ".format(c/100), end='')
                print("{0:.2f} ".format(b/100), end='')
                print("{0:.2f}".format(a/100))